cli/js/40_testing.js

```diff
@@ -145,25 +145,38 @@ function assertOps(fn) {
   return async function asyncOpSanitizer(desc) {
     const pre = core.metrics();
     const preTraces = new Map(core.opCallTraces);
+
     try {
       const innerResult = await fn(desc);
       if (innerResult) return innerResult;
-    } finally {
+    } catch (e) {
       // Defer until next event loop turn - that way timeouts and intervals
       // cleared can actually be removed from resource table, otherwise
       // false positives may occur (https://github.com/denoland/deno/issues/4591)
       await opSanitizerDelay();
+      throw e;
     }
-    const post = core.metrics();
-    const postTraces = new Map(core.opCallTraces);
 
+    let post = core.metrics();
     // We're checking diff because one might spawn HTTP server in the background
     // that will be a pending async op before test starts.
-    const dispatchedDiff = post.opsDispatchedAsync - pre.opsDispatchedAsync;
-    const completedDiff = post.opsCompletedAsync - pre.opsCompletedAsync;
+    let dispatchedDiff = post.opsDispatchedAsync - pre.opsDispatchedAsync;
+    let completedDiff = post.opsCompletedAsync - pre.opsCompletedAsync;
+
+    if (dispatchedDiff !== completedDiff) {
+      // Defer until next event loop turn - that way timeouts and intervals
+      // cleared can actually be removed from resource table, otherwise
+      // false positives may occur (https://github.com/denoland/deno/issues/4591)
+      await opSanitizerDelay();
+      post = core.metrics();
+      dispatchedDiff = post.opsDispatchedAsync - pre.opsDispatchedAsync;
+      completedDiff = post.opsCompletedAsync - pre.opsCompletedAsync;
+    }
```

Comment:
I think this could result in a false negative?

```ts
let counter = setTimeout(() => console.log("never", 1000000);

Deno.test(() => {
  clearTimeout(counter);
});
```

Reply:
Arguably an edge case - but we could fix this by checking `core.resources()` for any `timer` resources instead of `dispatchedDiff !== completedDiff`. If there are none, we can skip `opSanitizerDelay`.

Maybe also not worth fixing - just pointing it out :)

Reply:
Good point, there are two more points worth noting:
- I'd like to add timeouts to tests (with default one being 60s) which would render this PR useless, because every test would have a timeout - ideally we could rearchitect timeouts to be cheaper and not go through macrotask callback (that would fix the Async Local Storage problem as well)
- Ops related to worker threads suffer from the same problem - I need to dig into that problem and see if we can fix how they operate.

Reply:
I realized that this can also result in a false positive (symptoms would be similar to #20498). Unlike false negatives, false positives are unacceptable, so we will have to special case some ops irregardless of the diff count as per my previous comment https://github.com/denoland/deno/pull/20082#discussion_r1287608633.